### ISSUE TITLE
`renderNotFound()` looses state set by middleware

### ISSUE THREAD

**Issue by timharek** [denoland/fresh#1703]
An example:

```typescript
// routes/page.tsx

import { HandlerContext, Handlers, PageProps } from "$fresh/server.ts";
import { getContentFromShowere } from "./somewhere.ts";
import { State } from "./_middleware.ts";

interface Props {
   content: string;
}

export const handler: Handlers<Props, State> = {
  async GET(_req: Request, ctx: HandlerContext) {
    const content = await getContentFromSomewhere();
    if (!content) {
        return await ctx.renderNotFound();
    }
    return await ctx.render({ content });
  },
};

export default function Page(props: PageProps<Props>) {
  return <div>{props.content}</div>;
}
```

```typescript
// routes/_middleware.ts
import { MiddlewareHandlerContext } from "$fresh/server.ts";

export interface State {
  data: string;
}

export async function handler(
  req: Request,
  ctx: MiddlewareHandlerContext<State>
) {
  ctx.state.data = "myData";
  const resp = await ctx.next();
  resp.headers.set("server", "fresh server");
  return resp;
}
```

```typescript
// routes/_app.tsx
import { AppProps } from "$fresh/server.ts";
import { State } from "./_middleware.ts";

export default function App({ Component, state }: AppProps<State>) {
  return (
    <html>
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>My Fresh app</title>
        <meta name="description" content={state.data} />
      </head>
      <body>
        <Component />
      </body>
    </html>
  );
}
```

If `routes/page.tsx`'s function would return `null` or `undefined` and `ctx.renderNotFound()` would be called, then Fresh will crash because it's unable to retrieve the state set in the Middleware-handler. It's possible to fix this by passing the `ctx.state` in the `ctx.renderNotFound()` like this: `ctx.renderNotFound({ ...ctx.state })` and it will work. But it throws an error that `Props` are missing.

I see two possible solutions:

1. `renderNotFound()` is able to retrieve state set in the Middleware-handler.
2. We state from the routes handler (`ctx.renderNotFound({ ...ctx.state })`), and the handler props are not required, only the `State` ones.

A temporary solution I have used now is to make the Props optional in the handler and then make them required in Route-function.

**Comment by deer** [comment]
Are you sure? I can't reproduce this. Here is a more usable reproduction, based on the fresh source code and what you provided: https://github.com/deer/fresh/commit/ffbbf845d4bac01f333b439c41df660649a4f2d7

Have I misunderstood something? When I set a breakpoint in `_app.tsx` and `_404.tsx` I get access to the state just fine.

Additionally https://github.com/denoland/fresh/pull/1579 was recently completed, which makes it clear that 404s have access to state. #1579 even added a test for this case.

**Comment by timharek** [comment]
I created a example-project to demostrate it more properly and realized that `_app.tsx` has props called `data` and `state`. I think I have mixed them up, and that my projects have used `data` instead of `state`. And I believe the reason for the mix up is that I passed in the `State` interface as generic for `data`

So it looks like a user error from my side.

Sorry about that, I'm closing the issue.
